compiler/rustc_ty_utils/src/ty.rs

```diff
@@ -214,8 +220,36 @@ fn param_env(tcx: TyCtxt<'_>, def_id: DefId) -> ty::ParamEnv<'_> {
         _ => hir::CRATE_HIR_ID,
     };
 
-    let cause = traits::ObligationCause::misc(tcx.def_span(def_id), body_id);
-    traits::normalize_param_env_or_error(tcx, unnormalized_env, cause)
+    let span = tcx.def_span(def_id);
+
+    if is_fn {
```

Comment:
I'm skeptical that all this extra code is worthwhile... since all it's here for is to deduplicate a diagnostic that's still going to end up being an error. Especially because `normalize_param_env_or_error` is very delicate logic.

Reply:
Thanks for pointing it out. If that isn't worthwhile, could you take a look at the first commit,  which is just a little optimization of param_env, and I hope it's acceptable?

Reply:
Are you sure this is an optimization in reality? Maybe you should put it up so we can do a perf run on it. 

I am still not confident that the added complexity is worth it.